neovim: include lua51 headers to build properly

wryfi opened the pull request:
#### Description
<!-- Note: it is best to make pull requests from a branch rather than from master -->

Fixes neovim builds. _See_ [#70420](https://trac.macports.org/ticket/70420#comment:9) for full description.

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 14.6 23G80 arm64
Xcode 15.4 15F31d


###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

wryfi commented:
Note that I'm not a CMake expert, so I'm not certain that `target_include_directories(nlua0 AFTER PUBLIC /opt/local/include/lua5.1)` is strictly correct; but it does the job.

reneeotten commented on editors/neovim/files/patch-nlua0-includes.diff:
```diff
@@ -0,0 +1,12 @@
+   target_include_directories(main_lib SYSTEM BEFORE INTERFACE ${LUAJIT_INCLUDE_DIR})
+   target_link_libraries(main_lib INTERFACE ${LUAJIT_LIBRARY})
+   target_include_directories(nlua0 SYSTEM BEFORE PUBLIC ${LUAJIT_INCLUDE_DIR})
++  target_include_directories(nlua0 AFTER PUBLIC /opt/local/include/lua5.1)
```
don't hardcode "/opt/local" as user might install in a different ${prefix}. Not sure if there is a suitable CMake variable already available to use here. Otherwise the typical solution is to patch in a placeholder and then do a "reinplace" to put in "${prefix}" in the "post-patch" phase. There are many examples of ports doing similar thing you can check for inspiration.

reneeotten reviewed (commented):
Thanks @wryfi the patch isn't completely correct - I've left a comment. 

Also, please squash your commits and actually follow the guidelines for the commit message.

l2dy commented on editors/neovim/Portfile:
```diff
@@ -48,7 +48,7 @@ cmake.build_type        Release
 
 # Building parsers is normally an extra step, see https://github.com/neovim/neovim/issues/29042
-patchfiles              embed-parsers-build.diff
+patchfiles              embed-parsers-build.diff  patch-nlua0-includes.diff
```
Usually we put each patch file on their own line, like so:

```suggestion
patchfiles              embed-parsers-build.diff \
                        patch-nlua0-includes.diff
```

l2dy commented on editors/neovim/Portfile:
```diff
@@ -48,7 +48,13 @@ cmake.build_type        Release
+                        patch-lua51-includes.diff
+
+post-patch {
+    reinplace "s|@PREFIX@|${prefix}|g" \
+    ${worksrcpath}/CMakeLists.txt
```
Line 56 should be indented more than line 55. You could also merge the two lines like so:

```suggestion
    reinplace "s|@PREFIX@|${prefix}|g" ${worksrcpath}/CMakeLists.txt
```

l2dy reviewed (commented):
The commit message should start with `neovim:`, reflecting that you are fixing the neovim port.

Please also update the pull request title to match your commit message.

wryfi commented:
@reneeotten @l2dy thanks for your feedback on my first MacPorts PR! 

I fixed the issues you raised, and I also moved the fix to the top level of the cmake project, as I discovered (not surprisingly, I guess) that there are other instances of `#include <lua.h>` throughout the neovim sources.

l2dy commented on editors/neovim/files/patch-lua51-includes.diff:
```diff
@@ -0,0 +1,14 @@
+ # Version should match the tested CMAKE_URL in .githuworkflows/build.yml.
+ cmake_minimum_required(VERSION 3.13)
+
++include_directories(BEFORE @PREFIX@/include/lua5.1)
```
If neovim uses `luajit` as its runtime, should `/opt/local/include/luajit-2.1` be used here instead?

wryfi replied:
No, the issue is with the [nlua0](https://github.com/neovim/neovim/blob/8df6736ca14d09f87cf0a8486758ac5708819434/src/nlua0.c) C library that neovim builds, and then subsequently uses with luajit. `nlua0` is getting built against the lua 5.3 headers from the `lua` port, and then luajit chokes on `nlua0` because it is compiled against the wrong version of lua.

(I documented all of my findings extensively in [#70420](https://trac.macports.org/ticket/70420).

l2dy commented on editors/neovim/Portfile:
```diff
@@ -48,7 +48,12 @@ cmake.build_type        Release
 
 # Building parsers is normally an extra step, see https://github.com/neovim/neovim/issues/29042
-patchfiles              embed-parsers-build.diff
+patchfiles              embed-parsers-build.diff  \
```
Extra space could be removed.

```suggestion
patchfiles              embed-parsers-build.diff \
```

wryfi commented:
I also discovered that my fix won't work if someone tries to build neovim with the `PREFER_LUA` option, which will force neovim to build with lua 5.1 instead of luajit. 

I don't know if or why anyone would want to do this, but I have a fix for that too if it seems worth adding. In theory it would be relevant if MacPorts supports any platforms where luajit doesn't fly. But seems most likely to just become another patch to maintain for little or no value.

wryfi commented:
Incidentally `sudo port -vst install` fails from master (as well as my branch) on unpacking the source tarball. I checked the box in the PR template because it was already broken in master.

l2dy commented:
Merged. Thank you!
